cardano-crypto-class/src/Cardano/Crypto/DSIGN/EcdsaSecp256k1.hs

```diff
@@ -55,9 +81,198 @@ import Cardano.Crypto.DSIGN.Class (
+  secpCtxPtr,
+  secpEcdsaSign,
+  secpEcdsaVerify,
+  secpEcdsaSignatureSerializeCompact,
+  secpEcPubkeySerialize,
+  secpEcCompressed,
+  secpEcdsaSignatureParseCompact,
+  secpEcPubkeyParse,
+  )
+
+-- | As ECDSA signatures on the SECP256k1 curve sign 32-byte hashes, rather than
+-- whole messages, we provide a helper (opaque) newtype to ensure that the size
+-- of the input for signing and verification is strictly bounded.
+newtype MessageHash = MH (PinnedSizedBytes 32)
+  deriving Eq via (PinnedSizedBytes 32)
+  deriving stock Show
+
+-- | Take a blob of bytes (which is presumed to be a 32-byte hash), verify its
+-- length, and package it into a 'MessageHash' if that length is exactly 32.
+toMessageHash :: ByteString -> Maybe MessageHash
+toMessageHash bs = MH <$> psbFromByteStringCheck bs
+
+-- | Turn a 'MessageHash' into its bytes without a length marker.
+fromMessageHash :: MessageHash -> ByteString
+fromMessageHash (MH psb) = psbToByteString psb
+
+-- | A helper to use with the 'HashAlgorithm' API, as this can ensure sizing.
+hashAndPack :: forall (h :: Type) . 
+  (HashAlgorithm h, SizeHash h ~ 32) => 
+  Proxy h -> ByteString -> MessageHash
+hashAndPack p bs = case psbFromByteStringCheck . digest p $ bs of 
+  Nothing -> error $ "hashAndPack: unexpected mismatch of guaranteed hash length\n" <>
+                     "Please report this, it's a bug!"
+  Just psb -> MH psb
 
 data EcdsaSecp256k1DSIGN
 
+instance DSIGNAlgorithm EcdsaSecp256k1DSIGN where
+    type SeedSizeDSIGN EcdsaSecp256k1DSIGN = SECP256K1_ECDSA_PRIVKEY_BYTES
+    type SizeSigDSIGN EcdsaSecp256k1DSIGN = SECP256K1_ECDSA_SIGNATURE_BYTES
+    type SizeSignKeyDSIGN EcdsaSecp256k1DSIGN = SECP256K1_ECDSA_PRIVKEY_BYTES
+    type SizeVerKeyDSIGN EcdsaSecp256k1DSIGN = SECP256K1_ECDSA_PUBKEY_BYTES
+    type Signable EcdsaSecp256k1DSIGN = ((~) MessageHash)
+    newtype VerKeyDSIGN EcdsaSecp256k1DSIGN = 
+      VerKeyEcdsaSecp256k1 (PinnedSizedBytes SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL)
+      deriving newtype (Eq, NFData)
+      deriving stock (Show, Generic)
+      deriving anyclass (NoThunks)
+    newtype SignKeyDSIGN EcdsaSecp256k1DSIGN = 
+      SignKeyEcdsaSecp256k1 (PinnedSizedBytes SECP256K1_ECDSA_PRIVKEY_BYTES)
+      deriving newtype (Eq, NFData)
+      deriving stock (Show, Generic)
+      deriving anyclass (NoThunks)
+    newtype SigDSIGN EcdsaSecp256k1DSIGN = 
+      SigEcdsaSecp256k1 (PinnedSizedBytes SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL)
+      deriving newtype (Eq, NFData)
+      deriving stock (Show, Generic)
+      deriving anyclass (NoThunks)
+    algorithmNameDSIGN _ = "ecdsa-secp256k1"
+    {-# NOINLINE deriveVerKeyDSIGN #-}
+    deriveVerKeyDSIGN (SignKeyEcdsaSecp256k1 skBytes) = 
+      VerKeyEcdsaSecp256k1 <$> unsafeDupablePerformIO . psbUseAsSizedPtr skBytes $ 
+        \skp -> psbCreateSized $ \vkp -> do
+          res <- secpEcPubkeyCreate secpCtxPtr vkp skp
+          when (res /= 1) (error "deriveVerKeyDSIGN: Failed to derive pubkey")
```

Comment:
I think the key question is 'do we want this to be catchable'? The only way this could fail is either failure to meet a precondition of the call (which we ensure won't happen) or an internal problem of the `secp256k1` library itself (which we can't do anything about anyway). Catching in the first case is pointless (it shouldn't happen except when we have a bug), and in the second it's quite questionable (what _can_ you do if the underlying C fails you other than blow up?).

Reply:
I tend to agree with @kozross that this is not a recoverable exception, so there is no point in using `throwIO` in such cases.

Reply:
Well, when we talk about recoverable exceptions we have to consider that ultimately we're talking about a piece of software: the node. Do we want this to _crash the node_, or do we want this to potentially be caught by a top-level exception handler? (I guess people can catch `ErrorCall` but they usually don't) It seems unlikely to me that we _ever_ want to crash the top-level software.

Reply:
@michaelpj The node ought to catch all exceptions (i.e. `SomeException`), except the async ones, to avoid crashing. As you know there is plenty of code in Haskell that blows up in your face, `div`, `toEnum`, ... but we don't see applications catching all of those exception types individually eg. `ErrorCall`, `DivisionByZero`....

Most importantly, we are hiding here the fact that this code runs in IO with `unsafeDupablePerformIO`, therefore this function should be treated as a pure function with possibility of bottom in it. Recovering from exceptions that where thrown in pure code should be done only at the top level, as you mentioned, to prevent the application from crashing, but in no way it should be used for control flow, because order of evaluation in Haskell is non-deterministic, which is really bad for application control flow.

Now, if this was an IO action instead of an IO action disguised as a pure function, then I'd agree with you, we could add a special exception type and allow downstream users use it for control flow. But as it stands, I'd be really against it.

Reply:
Fine, I just also want to observe that `cardano-base` is inconsistent on this topic: https://github.com/input-output-hk/cardano-base/blob/1b4ac7bceb4f45554aa16d6cb3aa5d49ebb10c21/cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs#L69

But maybe that just means that that code is wrong!

Reply:
In this example we are actually getting the `errno`, which is something secp256k1 library does not exercise.

It would in fact make more sense to change it to:

```haskell
  error $ displayException $ errnoToIOError (contextDesc ++ ": " ++ cFunName) errno Nothing Nothing 
```

I hope no one is catching those `IOException`'s 

Overall I am not too concerned about this, because those are disaster like situation which will prevent the node to do any progress, because the only things can cause those crypto C libs to blow up after a node is running successfully are things like hardware failures or OOM.